rcl/include/rcl/publisher.h

```diff
@@ -20,7 +20,7 @@ extern "C"
 {
 #endif
 
-#include "rosidl_generator_c/message_type_support.h"
+#include "rosidl_generator_c/message_type_support_struct.h"
```

Comment:
I had a discussion about this with @esteve because I think it doesn't make sense to include publisher.h without committing to a middleware implementation. `message_type_support.h` is provided by individual rosidl type support packages, e.g.:

- https://github.com/ros2/rmw_opensplice/blob/master/rosidl_typesupport_opensplice_c/include/rosidl_typesupport_opensplice_c/impl/rosidl_generator_c/message_type_support.h
- https://github.com/ros2/rmw_connext/blob/master/rosidl_typesupport_connext_c/include/rosidl_typesupport_connext_c/impl/rosidl_generator_c/message_type_support.h

Where as `message_type_support_struct.h` is provided by the rmw implementation agnostic `rosidl_generator_c`.

So I'd say no, it's not quite right and instead changing this just worked around another issue which was that something you were building needed a specific type support but wasn't using its include directories.

Separately we discussed whether or not it made since to do this in publisher.h and/or subscription.h but not in things like node.h. It is absolutely necessary in node.h since that will use messages from rcl_interfaces and once you include it you will need to be tied to a specific rmw implementation, but technically publisher and subscription do not have this restriction, but they could in the future.